collector/src/bin/rustc-perf-collector/execute.rs

```diff
@@ -262,6 +258,265 @@ lazy_static! {
     };
 }
 
+pub trait Processor {
+    fn profiler(&self) -> Profiler;
+
+    fn process_output(&mut self, name: &str, cwd: &Path, build_kind: BuildKind, run_kind: RunKind,
+                      run_kind_str: &str, patch: Option<&Patch>, output: process::Output)
+                      -> Result<bool, Error>;
+
+    fn finish_build_kind(&mut self, _build_kind: BuildKind) {}
+}
+
+pub struct MeasureProcessor {
+    clean_stats: Vec<Vec<Stat>>,
+    nll_stats: Vec<Vec<Stat>>,
+    base_incr_stats: Vec<Vec<Stat>>,
+    clean_incr_stats: Vec<Vec<Stat>>,
+    patched_incr_stats: Vec<(Patch, Vec<Vec<Stat>>)>,
+
+    pub collected: CollectedBenchmark,
+}
+
+impl MeasureProcessor {
+    pub fn new(name: &str) -> Self {
+        // Check we have `perf` available.
+        let has_perf = Command::new("perf").output().is_ok();
+        assert!(has_perf);
+
+        MeasureProcessor {
+            clean_stats: Vec::new(),
+            nll_stats: Vec::new(),
+            base_incr_stats: Vec::new(),
+            clean_incr_stats: Vec::new(),
+            patched_incr_stats: Vec::new(),
+
+            collected: CollectedBenchmark {
+                name: name.to_string(),
+                runs: Vec::new(),
+            },
+        }
+    }
+}
+
+impl Processor for MeasureProcessor {
+    fn profiler(&self) -> Profiler {
+        Profiler::PerfStat
+    }
+
+    fn process_output(&mut self, _name: &str, _cwd: &Path, _build_kind: BuildKind,
+                      run_kind: RunKind, _run_kind_str: &str, patch: Option<&Patch>,
+                      output: process::Output) -> Result<bool, Error> {
+        match process_perf_stat_output(output) {
+            Ok(stats) => {
+                match run_kind {
+                    RunKind::Clean => { self.clean_stats.push(stats); }
+                    RunKind::Nll => { self.nll_stats.push(stats); }
+                    RunKind::BaseIncr => { self.base_incr_stats.push(stats); }
+                    RunKind::CleanIncr => { self.clean_incr_stats.push(stats); }
+                    RunKind::PatchedIncrs => {
+                        let patch = patch.unwrap();
+                        if let Some(mut entry) =
+                            self.patched_incr_stats.iter_mut().find(|s| &s.0 == patch)
+                        {
+                            entry.1.push(stats);
+                            return Ok(true);
+                        }
+                        self.patched_incr_stats.push((patch.clone(), vec![stats]));
+                    }
+                }
+                Ok(true)
+            }
+	    Err(DeserializeStatError::NoOutput(output)) => {
+		warn!(
+		    "failed to deserialize stats, retrying; output: {:?}",
+		    output
+		);
+                Ok(false)   // retry
```

Comment:
Could we change this bool to an enum? Not quite sure what to call it, but maybe for now just `Retry::{Yes, No}`. I guess to an extent it feels like we should be returning `Result<(), SomeError>` from this and retry would be indicated in the error enum/struct -- I think failure should allow this fairly easily.